### DISCUSSION TITLE
Automate user creation and role assignment.

### DISCUSSION THREAD

**Discussion by clevertech-lazaropj** [infinispan/infinispan#12691] [General]
Hello!

I'm new to Infinispan and I'm trying to integrate an Infinispan server with Keycloak using Kubernetes on GCP. I'm looking for a way to automate user creation and role assignment through a Kubernetes Job. I've tried using the CLI interface, but it requires interactive access to function correctly, so I can't execute the user grant role command directly in the Job. I've also read about batch CRs, but I would prefer to find a solution that doesn't require installing additional components first (like the Infinispan operator).

**Comment by ryanemerson** [comment]
I would strongly recommend install the Infinispan Operator, as it's purpose is to simplify these very tasks :slightly_smiling_face:. There's a Keycloak specific guide that explains how it can help you automate such tasks.

Alternatively, we do provide a helm chart that allows you to provide a CLI batch script via a Kubernetes secret using the `deploy.security.secretName` field:

- https://infinispan.org/docs/helm-chart/main/helm-chart.html#adding-multiple-credentials_configuring-authentication
- https://github.com/infinispan/infinispan-helm-charts/

**Reply by clevertech-lazaropj** [reply]
Thank you for your response!

I'll definitely look into the Infinispan Operator in more detail.

Regarding the CLI batch script, I've only seen examples for user creation (including group creation with the -g flag). In this setup, if I were to execute a grant command, would it work? I'm asking because I can't execute the grant command without connecting to an Infinispan server first, but I can create users without connecting to a server. This is actually the issue I'm facing right now: I can create users using the CLI, but I cannot assign roles without first running the connect command, which unfortunately only works through an interactive session.

**Reply by tristantarrant** [reply]
The reason why `grant` doesn't work offline is because it manipulates how users/groups are mapped to roles. The mapping information is stored in a replicated cache (an implementation detail of the cluster role mapper).

**Reply by clevertech-lazaropj** [reply]
Thanks! Another thing. I don't understand how the CLI batch script in the secrets.yaml is actually going to be executed inside my container by the CLI. I'm having a hard time figuring out how the secrets goes together with my deployment.yaml file, which actually deploys my contatiner. If necessary, i can post a snippet of the code in here.

**Reply by ryanemerson** [reply]
In the helm chart, the `identities-batch` file that you create in your secret is mounted to `/etc/security/identities-batch`. This file path is then set as the value for the `IDENTITIES_BATCH` Environment variable which the Infinispan image uses as an argument for an offline CLI instance.

https://github.com/infinispan/infinispan-images/blob/main/modules/org/infinispan/runtime/added/bin/launch.sh#L103-#L106
